### ISSUE TITLE
[Question] IBC training and partial crossgen on Linux

### ISSUE THREAD

**Issue by swift-kim** [dotnet/coreclr#25787]
I'm trying to generate optimized (partial-ngened) assemblies against user applications using IBC data on Linux system (armel Tizen). The system is running the latest coreclr runtime and crossgen built from source. I have no other internal tools like `ibcmerge.exe`.

I have created instrumented images of some target assemblies (including system assemblies), and obtained training data (.ibc files) by running different applications several times. Here is a [sample](https://github.com/dotnet/coreclr/files/3409654/ibc-new.zip). As far as I see in HexViewer (I found no other ways to read the files), the files have similar structures to those in _optimization.linux-x64.ibc.corefx_ therefore they look quite valid.

I have run crossgen again with the IBC data as input and passing `PartialNGen`/`UseIBCFile` as true, and in fact it could generate some kind of optimized images for the assemblies. However, the profiling result shows they are optimized for only the first application I have run for training, but not exactly for other applications.

**TL;DR** - Is it able to train partial-ngened assemlibes for mutiple scenarios/runs based purely on crossgen?

Any comments will help a lot. Thank you.

**Comment by AaronRobinsonMSFT** [comment]
cc @briansull

**Comment by jkotas** [comment]
> Is it able to train partial-ngened assemlibes for mutiple scenarios/runs based purely on crossgen?

It should be possible, but it may be broken currently because of it is not used by anything. Could you please try to debug through crossgen to see why it is not using all the training data? `ZapImage::CompileHotRegion()` method is a good place to start.

**Comment by briansull** [comment]
I also need the matching dll's in order  to examine or doi anything with these .ibc files:

> Here is a sample.

**Comment by swift-kim** [comment]
@briansull 
Thank you for having a look. [This zip](https://github.com/dotnet/coreclr/files/3415298/ibc-dlls.zip) file contains the corresponding DLLs. (Note: The command line section starting with `/usr/bin/dotnet-launcher ...` will look super scary, but it's normal.)

@jkotas 
I've taken a glance at `ZapImage::parseProfileData()` which initially reads data into `m_profileDataSections` but as far as I can see, it deals only with a single `CORBBTPROF_SECTION_TABLE_HEADER` (therefore ignoring the rest that there may exist) in an ibc file. Am I getting something wrong?

**Comment by jkotas** [comment]
I think it should be fine to fix `parseProfileData` to deal with multiple `CORBBTPROF_SECTION_TABLE_HEADER` sections.

**Comment by swift-kim** [comment]
@briansull
Could I ask what `ibcmerge.exe` is normally supposed to do with ibc files and DLLs? Is it responsible for combining and reorganizing profiling data from multiple scenarios in a single/multiple ibc files in general? If so, which one do you think is responsible for the job in a scenario where we don't have `ibcmerge.exe`, EE or crossgen?

**Comment by briansull** [comment]
I examined your .ibc data files and they appear to be corrupt as the IbcMerge tool has trouble parsing.them.  It fails at the point of reading the DateTime information in the IBC Scenario Data.
When I added a try.catch to proceed past this point the data structure is still corrupted as the 
Command line length information is wrong.

This is an area that was updated to get things working for LInux in CoreCLR 3.0, so it is possible that you don't have the correct set ofchanges here or that you version of Linux is returning different values.

```
System.ArgumentOutOfRangeException: 'Not a valid Win32 FileTime.
Parameter name: fileTime'
```

**Comment by briansull** [comment]
I don't believe it is possible for use to use the  IBC data to optimize your assemblies with using the Ibcmerge.exe tool to pre-process the profile data before attaching it to the assembly as a resource.

**Comment by swift-kim** [comment]
@briansull 
I actually made another change in `ProfileDataAllocateScenarioInfo()` and probably it's breaking the IbcMerge parser. What I did was adding some `ALIGN_UP()`s in allocation of profile buffers to avoid SIGBUS on armel platform (due to an unaligned memory access), but I didn't update the sizes in structures (`cName`, `cCmdLine`, `cSystemInfo`) that would be expected by the parser.

The following patch I made to the current master would fix the issue. Please check [**this file**](https://github.com/dotnet/coreclr/files/3425149/ibc-weather.zip) which contains regenerated ibc files with fix applied.

```patch
diff --git a/src/vm/ceeload.cpp b/src/vm/ceeload.cpp
index ae92074952..f8fa6c876f 100644
--- a/src/vm/ceeload.cpp
+++ b/src/vm/ceeload.cpp
@@ -11889,16 +11889,16 @@ static void ProfileDataAllocateScenarioInfo(ProfileEmitter * pEmitter, LPCSTR sc
             pCmdLine = GetCommandLineW();
         }
 
-        S_SIZE_T cCmdLine = S_SIZE_T(wcslen(pCmdLine));
-        cCmdLine += 1;
+        S_SIZE_T cCmdLine = S_SIZE_T((wcslen(pCmdLine)+2) & ~1); // Adds 1 (if odd), or 2 (if even).
+        // cCmdLine += 1;
         if (cCmdLine.IsOverflow())
         {
             ThrowHR(COR_E_OVERFLOW);
         }
         
         LPCWSTR  pSystemInfo = W("<machine,OS>");
-        S_SIZE_T cSystemInfo = S_SIZE_T(wcslen(pSystemInfo));
-        cSystemInfo += 1;
+        S_SIZE_T cSystemInfo = S_SIZE_T((wcslen(pSystemInfo)+2) & ~1);
+        // cSystemInfo += 1;
         if (cSystemInfo.IsOverflow())
         {
             ThrowHR(COR_E_OVERFLOW);
@@ -11911,8 +11911,8 @@ static void ProfileDataAllocateScenarioInfo(ProfileEmitter * pEmitter, LPCSTR sc
         GetBasename(pCmdLine, &scenarioName[0], 256);
         
         LPCWSTR  pName      = &scenarioName[0];
-        S_SIZE_T cName      = S_SIZE_T(wcslen(pName));
-        cName += 1;
+        S_SIZE_T cName      = S_SIZE_T((wcslen(pName)+2) & ~1);
+        // cName += 1;
         if (cName.IsOverflow())
         {
             ThrowHR(COR_E_OVERFLOW);
```

If the corruption has still not been resolved, or I you find a better way to fix the aligning issue, please let me know. Also, I want to make it clear that our final goal is to optimize user applications without relying on IbcMerge (and I wonder if you agree with making such changes in coreclr that would not require IbcMerge for general use cases on Linux.)

**Comment by briansull** [comment]
I was able to use the ibcmerge.exe tool to examine your new ibc data, it looks fine now:

You ran 6 scenarios:  
Here is an example from **Tizen.Applications.Common.ibc.xml**

```
ï»¿<?xml version="1.0" encoding="utf-8"?>
<ProfileData version="2" mvid="ccab659c-f3b6-4b39-b580-dde16b086dfb">
  <ScenarioSection totalNumRuns="6" scenarioCount="6">
    <Scenario ordinal="1" scenario="A" mask="0x0020" priority="0" numRuns="1" name="/usr/bin/dotnet-launcher">
      <Run runTime="2/7/2016 1:19:20 AM" cmdLine="/usr/bin/dotnet-launcher /opt/usr/globalapps/org.tizen.example.Alarm/bin/Alarm.dll `zaybxcwdveuftgsh` __APP_SVC_OP_TYPE__ UgAAAAEEAAAUAAAAX19BUFBfU1ZDX09QX1RZUEVfXwAuAAAAaHR0cDovL3RpemVuLm9yZy9hcHBjb250cm9sL29wZXJhdGlvbi9kZWZhdWx0AA== __AUL_STARTTIME__ NAAAAAEEAAASAAAAX19BVUxfU1RBUlRUSU1FX18AEgAAADE0NTQ4MzY3MzkvODg0NzczAA== __AUL_TARGET_UID__ KAAAAAEEAAATAAAAX19BVUxfVEFSR0VUX1VJRF9fAAUAAAA1MDAxAA== __APP_SVC_PKG_NAME__ PQAAAAEEAAAVAAAAX19BUFBfU1ZDX1BLR19OQU1FX18AGAAAAG9yZy50aXplbi5leGFtcGxlLkFsYXJtAA== __AUL_CALLER_PID__ KAAAAAEEAAATAAAAX19BVUxfQ0FMTEVSX1BJRF9fAAUAAAAyNjk4AA== __AUL_CALLER_UID__ KAAAAAEEAAATAAAAX19BVUxfQ0FMTEVSX1VJRF9fAAUAAAA1MDAxAA== __AUL_CALLER_APPID__ NgAAAAEEAAAVAAAAX19BVUxfQ0FMTEVSX0FQUElEX18AEQAAAG9yZy50aXplbi53LWhvbWUA __AUL_HWACC__ IgAAAAEEAAAOAAAAX19BVUxfSFdBQ0NfXwAEAAAAU1lTAA== __AUL_ROOT_PATH__ TgAAAAEEAAASAAAAX19BVUxfUk9PVF9QQVRIX18ALAAAAC9vcHQvdXNyL2dsb2JhbGFwcHMvb3JnLnRpemVuLmV4YW1wbGUuQWxhcm0A __AUL_INTERNAL_POOL__ LAAAAAEEAAAWAAAAX19BVUxfSU5URVJOQUxfUE9PTF9fAAYAAABmYWxzZQA= __AUL_API_VERSION__ JgAAAAEEAAAUAAAAX19BVUxfQVBJX1ZFUlNJT05fXwACAAAANAA= __AUL_PROFILE__ KQAAAAEEAAAQAAAAX19BVUxfUFJPRklMRV9fAAkAAAB3ZWFyYWJsZQA= __AUL_IS_GLOBAL__ JwAAAAEEAAASAAAAX19BVUxfSVNfR0xPQkFMX18ABQAAAHRydWUA __AUL_INSTALLED_STORAGE__ MwAAAAEEAAAaAAAAX19BVUxfSU5TVEFMTEVEX1NUT1JBR0VfXwAJAAAAaW50ZXJuYWwA" systemInfo="&lt;machine,OS&gt;" mvid="ccab659c-f3b6-4b39-b580-dde16b086dfb" />
    </Scenario>
    <Scenario ordinal="2" scenario="B" mask="0x0010" priority="0" numRuns="1" name="/usr/bin/dotnet-launcher">
      <Run runTime="2/7/2016 1:19:56 AM" cmdLine="/usr/bin/dotnet-launcher /opt/usr/globalapps/org.tizen.example.Calculator/bin/Calculator.dll `zaybxcwdveuftgsh` __APP_SVC_OP_TYPE__ UgAAAAEEAAAUAAAAX19BUFBfU1ZDX09QX1RZUEVfXwAuAAAAaHR0cDovL3RpemVuLm9yZy9hcHBjb250cm9sL29wZXJhdGlvbi9kZWZhdWx0AA== __AUL_STARTTIME__ NAAAAAEEAAASAAAAX19BVUxfU1RBUlRUSU1FX18AEgAAADE0NTQ4MzY3NzQvMjAzMDI5AA== __AUL_TARGET_UID__ KAAAAAEEAAATAAAAX19BVUxfVEFSR0VUX1VJRF9fAAUAAAA1MDAxAA== __APP_SVC_PKG_NAME__ QgAAAAEEAAAVAAAAX19BUFBfU1ZDX1BLR19OQU1FX18AHQAAAG9yZy50aXplbi5leGFtcGxlLkNhbGN1bGF0b3IA __AUL_CALLER_PID__ KAAAAAEEAAATAAAAX19BVUxfQ0FMTEVSX1BJRF9fAAUAAAAyNjk4AA== __AUL_CALLER_UID__ KAAAAAEEAAATAAAAX19BVUxfQ0FMTEVSX1VJRF9fAAUAAAA1MDAxAA== __AUL_CALLER_APPID__ NgAAAAEEAAAVAAAAX19BVUxfQ0FMTEVSX0FQUElEX18AEQAAAG9yZy50aXplbi53LWhvbWUA __AUL_HWACC__ IgAAAAEEAAAOAAAAX19BVUxfSFdBQ0NfXwAEAAAAU1lTAA== __AUL_ROOT_PATH__ UwAAAAEEAAASAAAAX19BVUxfUk9PVF9QQVRIX18AMQAAAC9vcHQvdXNyL2dsb2JhbGFwcHMvb3JnLnRpemVuLmV4YW1wbGUuQ2FsY3VsYXRvcgA= __AUL_INTERNAL_POOL__ LAAAAAEEAAAWAAAAX19BVUxfSU5URVJOQUxfUE9PTF9fAAYAAABmYWxzZQA= __AUL_API_VERSION__ JgAAAAEEAAAUAAAAX19BVUxfQVBJX1ZFUlNJT05fXwACAAAANAA= __AUL_PROFILE__ KQAAAAEEAAAQAAAAX19BVUxfUFJPRklMRV9fAAkAAAB3ZWFyYWJsZQA= __AUL_IS_GLOBAL__ JwAAAAEEAAASAAAAX19BVUxfSVNfR0xPQkFMX18ABQAAAHRydWUA __AUL_INSTALLED_STORAGE__ MwAAAAEEAAAaAAAAX19BVUxfSU5TVEFMTEVEX1NUT1JBR0VfXwAJAAAAaW50ZXJuYWwA" systemInfo="&lt;machine,OS&gt;" mvid="ccab659c-f3b6-4b39-b580-dde16b086dfb" />
    </Scenario>
    <Scenario ordinal="3" scenario="C" mask="0x0008" priority="0" numRuns="1" name="/usr/bin/dotnet-launcher">
      <Run runTime="2/7/2016 1:20:22 AM" cmdLine="/usr/bin/dotnet-launcher /opt/usr/globalapps/org.tizen.example.HeartRateMonitor/bin/HeartRateMonitor.dll `zaybxcwdveuftgsh` __APP_SVC_OP_TYPE__ UgAAAAEEAAAUAAAAX19BUFBfU1ZDX09QX1RZUEVfXwAuAAAAaHR0cDovL3RpemVuLm9yZy9hcHBjb250cm9sL29wZXJhdGlvbi9kZWZhdWx0AA== __AUL_STARTTIME__ NAAAAAEEAAASAAAAX19BVUxfU1RBUlRUSU1FX18AEgAAADE0NTQ4MzY4MDUvNjk2NDkwAA== __AUL_TARGET_UID__ KAAAAAEEAAATAAAAX19BVUxfVEFSR0VUX1VJRF9fAAUAAAA1MDAxAA== __APP_SVC_PKG_NAME__ SAAAAAEEAAAVAAAAX19BUFBfU1ZDX1BLR19OQU1FX18AIwAAAG9yZy50aXplbi5leGFtcGxlLkhlYXJ0UmF0ZU1vbml0b3IA __AUL_CALLER_PID__ KAAAAAEEAAATAAAAX19BVUxfQ0FMTEVSX1BJRF9fAAUAAAAyNjk4AA== __AUL_CALLER_UID__ KAAAAAEEAAATAAAAX19BVUxfQ0FMTEVSX1VJRF9fAAUAAAA1MDAxAA== __AUL_CALLER_APPID__ NgAAAAEEAAAVAAAAX19BVUxfQ0FMTEVSX0FQUElEX18AEQAAAG9yZy50aXplbi53LWhvbWUA __AUL_HWACC__ IgAAAAEEAAAOAAAAX19BVUxfSFdBQ0NfXwAEAAAAU1lTAA== __AUL_ROOT_PATH__ WQAAAAEEAAASAAAAX19BVUxfUk9PVF9QQVRIX18ANwAAAC9vcHQvdXNyL2dsb2JhbGFwcHMvb3JnLnRpemVuLmV4YW1wbGUuSGVhcnRSYXRlTW9uaXRvcgA= __AUL_INTERNAL_POOL__ LAAAAAEEAAAWAAAAX19BVUxfSU5URVJOQUxfUE9PTF9fAAYAAABmYWxzZQA= __AUL_API_VERSION__ JgAAAAEEAAAUAAAAX19BVUxfQVBJX1ZFUlNJT05fXwACAAAANAA= __AUL_PROFILE__ KQAAAAEEAAAQAAAAX19BVUxfUFJPRklMRV9fAAkAAAB3ZWFyYWJsZQA= __AUL_IS_GLOBAL__ JwAAAAEEAAASAAAAX19BVUxfSVNfR0xPQkFMX18ABQAAAHRydWUA __AUL_INSTALLED_STORAGE__ MwAAAAEEAAAaAAAAX19BVUxfSU5TVEFMTEVEX1NUT1JBR0VfXwAJAAAAaW50ZXJuYWwA" systemInfo="&lt;machine,OS&gt;" mvid="ccab659c-f3b6-4b39-b580-dde16b086dfb" />
    </Scenario>
    <Scenario ordinal="4" scenario="D" mask="0x0004" priority="0" numRuns="1" name="/usr/bin/dotnet-launcher">
      <Run runTime="2/7/2016 1:21:01 AM" cmdLine="/usr/bin/dotnet-launcher /opt/usr/globalapps/org.tizen.example.Net.VoiceMemo.Tizen.Wearable/bin/VoiceMemo.Tizen.Wearable.dll `zaybxcwdveuftgsh` __APP_SVC_OP_TYPE__ UgAAAAEEAAAUAAAAX19BUFBfU1ZDX09QX1RZUEVfXwAuAAAAaHR0cDovL3RpemVuLm9yZy9hcHBjb250cm9sL29wZXJhdGlvbi9kZWZhdWx0AA== __AUL_STARTTIME__ MwAAAAEEAAASAAAAX19BVUxfU1RBUlRUSU1FX18AEQAAADE0NTQ4MzY4MzQvMTYxMTIA __AUL_TARGET_UID__ KAAAAAEEAAATAAAAX19BVUxfVEFSR0VUX1VJRF9fAAUAAAA1MDAxAA== __APP_SVC_PKG_NAME__ VAAAAAEEAAAVAAAAX19BUFBfU1ZDX1BLR19OQU1FX18ALwAAAG9yZy50aXplbi5leGFtcGxlLk5ldC5Wb2ljZU1lbW8uVGl6ZW4uV2VhcmFibGUA __AUL_CALLER_PID__ KAAAAAEEAAATAAAAX19BVUxfQ0FMTEVSX1BJRF9fAAUAAAAyNjk4AA== __AUL_CALLER_UID__ KAAAAAEEAAATAAAAX19BVUxfQ0FMTEVSX1VJRF9fAAUAAAA1MDAxAA== __AUL_CALLER_APPID__ NgAAAAEEAAAVAAAAX19BVUxfQ0FMTEVSX0FQUElEX18AEQAAAG9yZy50aXplbi53LWhvbWUA __AUL_ALLOWED_BG__ LgAAAAEEAAATAAAAX19BVUxfQUxMT1dFRF9CR19fAAsAAABBTExPV0VEX0JHAA== __AUL_HWACC__ IgAAAAEEAAAOAAAAX19BVUxfSFdBQ0NfXwAEAAAAU1lTAA== __AUL_ROOT_PATH__ ZQAAAAEEAAASAAAAX19BVUxfUk9PVF9QQVRIX18AQwAAAC9vcHQvdXNyL2dsb2JhbGFwcHMvb3JnLnRpemVuLmV4YW1wbGUuTmV0LlZvaWNlTWVtby5UaXplbi5XZWFyYWJsZQA= __AUL_INTERNAL_POOL__ LAAAAAEEAAAWAAAAX19BVUxfSU5URVJOQUxfUE9PTF9fAAYAAABmYWxzZQA= __AUL_API_VERSION__ JgAAAAEEAAAUAAAAX19BVUxfQVBJX1ZFUlNJT05fXwACAAAANAA= __AUL_PROFILE__ KQAAAAEEAAAQAAAAX19BVUxfUFJPRklMRV9fAAkAAAB3ZWFyYWJsZQA= __AUL_IS_GLOBAL__ JwAAAAEEAAASAAAAX19BVUxfSVNfR0xPQkFMX18ABQAAAHRydWUA __AUL_INSTALLED_STORAGE__ MwAAAAEEAAAaAAAAX19BVUxfSU5TVEFMTEVEX1NUT1JBR0VfXwAJAAAAaW50ZXJuYWwA" systemInfo="&lt;machine,OS&gt;" mvid="ccab659c-f3b6-4b39-b580-dde16b086dfb" />
    </Scenario>
    <Scenario ordinal="5" scenario="E" mask="0x0002" priority="0" numRuns="1" name="/usr/bin/dotnet-launcher">
      <Run runTime="2/7/2016 1:21:39 AM" cmdLine="/usr/bin/dotnet-launcher /opt/usr/globalapps/org.tizen.example.Weather/bin/Weather.dll `zaybxcwdveuftgsh` __APP_SVC_OP_TYPE__ UgAAAAEEAAAUAAAAX19BUFBfU1ZDX09QX1RZUEVfXwAuAAAAaHR0cDovL3RpemVuLm9yZy9hcHBjb250cm9sL29wZXJhdGlvbi9kZWZhdWx0AA== __AUL_STARTTIME__ NAAAAAEEAAASAAAAX19BVUxfU1RBUlRUSU1FX18AEgAAADE0NTQ4MzY4NzAvOTQyMDg4AA== __AUL_TARGET_UID__ KAAAAAEEAAATAAAAX19BVUxfVEFSR0VUX1VJRF9fAAUAAAA1MDAxAA== __APP_SVC_PKG_NAME__ PwAAAAEEAAAVAAAAX19BUFBfU1ZDX1BLR19OQU1FX18AGgAAAG9yZy50aXplbi5leGFtcGxlLldlYXRoZXIA __AUL_CALLER_PID__ KAAAAAEEAAATAAAAX19BVUxfQ0FMTEVSX1BJRF9fAAUAAAAyNjk4AA== __AUL_CALLER_UID__ KAAAAAEEAAATAAAAX19BVUxfQ0FMTEVSX1VJRF9fAAUAAAA1MDAxAA== __AUL_CALLER_APPID__ NgAAAAEEAAAVAAAAX19BVUxfQ0FMTEVSX0FQUElEX18AEQAAAG9yZy50aXplbi53LWhvbWUA __AUL_HWACC__ IgAAAAEEAAAOAAAAX19BVUxfSFdBQ0NfXwAEAAAAU1lTAA== __AUL_ROOT_PATH__ UAAAAAEEAAASAAAAX19BVUxfUk9PVF9QQVRIX18ALgAAAC9vcHQvdXNyL2dsb2JhbGFwcHMvb3JnLnRpemVuLmV4YW1wbGUuV2VhdGhlcgA= __AUL_INTERNAL_POOL__ LAAAAAEEAAAWAAAAX19BVUxfSU5URVJOQUxfUE9PTF9fAAYAAABmYWxzZQA= __AUL_API_VERSION__ JgAAAAEEAAAUAAAAX19BVUxfQVBJX1ZFUlNJT05fXwACAAAANAA= __AUL_PROFILE__ KQAAAAEEAAAQAAAAX19BVUxfUFJPRklMRV9fAAkAAAB3ZWFyYWJsZQA= __AUL_IS_GLOBAL__ JwAAAAEEAAASAAAAX19BVUxfSVNfR0xPQkFMX18ABQAAAHRydWUA __AUL_INSTALLED_STORAGE__ MwAAAAEEAAAaAAAAX19BVUxfSU5TVEFMTEVEX1NUT1JBR0VfXwAJAAAAaW50ZXJuYWwA" systemInfo="&lt;machine,OS&gt;" mvid="ccab659c-f3b6-4b39-b580-dde16b086dfb" />
    </Scenario>
    <Scenario ordinal="6" scenario="F" mask="0x0001" priority="0" numRuns="1" name="/usr/bin/dotnet-launcher">
      <Run runTime="2/7/2016 1:22:10 AM" cmdLine="/usr/bin/dotnet-launcher /opt/usr/globalapps/org.tizen.example.XStopWatch/bin/XStopWatch.dll `zaybxcwdveuftgsh` __APP_SVC_OP_TYPE__ UgAAAAEEAAAUAAAAX19BUFBfU1ZDX09QX1RZUEVfXwAuAAAAaHR0cDovL3RpemVuLm9yZy9hcHBjb250cm9sL29wZXJhdGlvbi9kZWZhdWx0AA== __AUL_STARTTIME__ NAAAAAEEAAASAAAAX19BVUxfU1RBUlRUSU1FX18AEgAAADE0NTQ4MzY5MTEvNjMxMTE0AA== __AUL_TARGET_UID__ KAAAAAEEAAATAAAAX19BVUxfVEFSR0VUX1VJRF9fAAUAAAA1MDAxAA== __APP_SVC_PKG_NAME__ QgAAAAEEAAAVAAAAX19BUFBfU1ZDX1BLR19OQU1FX18AHQAAAG9yZy50aXplbi5leGFtcGxlLlhTdG9wV2F0Y2gA __AUL_CALLER_PID__ KAAAAAEEAAATAAAAX19BVUxfQ0FMTEVSX1BJRF9fAAUAAAAyNjk4AA== __AUL_CALLER_UID__ KAAAAAEEAAATAAAAX19BVUxfQ0FMTEVSX1VJRF9fAAUAAAA1MDAxAA== __AUL_CALLER_APPID__ NgAAAAEEAAAVAAAAX19BVUxfQ0FMTEVSX0FQUElEX18AEQAAAG9yZy50aXplbi53LWhvbWUA __AUL_HWACC__ IgAAAAEEAAAOAAAAX19BVUxfSFdBQ0NfXwAEAAAAU1lTAA== __AUL_ROOT_PATH__ UwAAAAEEAAASAAAAX19BVUxfUk9PVF9QQVRIX18AMQAAAC9vcHQvdXNyL2dsb2JhbGFwcHMvb3JnLnRpemVuLmV4YW1wbGUuWFN0b3BXYXRjaAA= __AUL_INTERNAL_POOL__ LAAAAAEEAAAWAAAAX19BVUxfSU5URVJOQUxfUE9PTF9fAAYAAABmYWxzZQA= __AUL_API_VERSION__ JgAAAAEEAAAUAAAAX19BVUxfQVBJX1ZFUlNJT05fXwACAAAANAA= __AUL_PROFILE__ KQAAAAEEAAAQAAAAX19BVUxfUFJPRklMRV9fAAkAAAB3ZWFyYWJsZQA= __AUL_IS_GLOBAL__ JwAAAAEEAAASAAAAX19BVUxfSVNfR0xPQkFMX18ABQAAAHRydWUA __AUL_INSTALLED_STORAGE__ MwAAAAEEAAAaAAAAX19BVUxfSU5TVEFMTEVEX1NUT1JBR0VfXwAJAAAAaW50ZXJuYWwA" systemInfo="&lt;machine,OS&gt;" mvid="ccab659c-f3b6-4b39-b580-dde16b086dfb" />
    </Scenario>
  </ScenarioSection>
  <BasicBlockInfo count="30">
    <Method token="0x06000003" ILSize="0x00000040" name="instance void Tizen.Applications.AppControl::.ctor(Tizen.Applications.SafeAppControlHandle)" count="5">
      <BasicBlock offset="0x000" count="6" />
      <BasicBlock offset="0x009" count="0" />
      <BasicBlock offset="0x014" count="6" />
      <BasicBlock offset="0x029" count="0" />
      <BasicBlock offset="0x03f" count="6" />
    </Method>
    <Method token="0x06000012" ILSize="0x00000020" name="instance Tiz
```

**Comment by swift-kim** [comment]
Related to #20496.

Commented by @adiaaida,
>This would be the solution until ibcmerge is cross-plat, if the decision to do that is ever made.

**Comment by briansull** [comment]
The IBCMerge tool has been ported to work on Linux systems, so IBCMerge is now is multi-platform.
But I'm not sure exactly what you are asking for here.   
It looks like you can now successfully generate IBC profile data when running your scenarios.

And in order to utilize partial ngen via crossgen the IbcMerge tool must be used to attach the profile data to each assembly that wants to utilize partial ngen.

**Comment by swift-kim** [comment]
@briansull I was to figure out the current availability and plans for partial ngen on Linux.

> in order to utilize partial ngen via crossgen the IbcMerge tool **must be used**

That answers the original question. Maybe I will need an access to the tool, or we should implement such funtion in crossgen to utilize multiple profile data (directly from ibc files) without the tool.

Thank you.

**Comment by swift-kim** [comment]
I'll close this issue now. Thanks for all your help.
